provider/image.go

```diff
@@ -201,8 +205,12 @@ func (p *dockerNativeProvider) dockerBuild(ctx context.Context,
 			}
 		}
 		// we push to the server declared in the program, using our auth configs from image build.
-		// if the program does not have a server declared, we will let the docker client error
-		pushAuthConfig = authConfigs[img.Registry.Server]
+		// if there is no servername in the registry, we attempt to build it from the fully qualified image name.
+		registryServer, err := getRegistryAddr(img.Registry.Server, img.Name)
+		if err != nil {
+			return "", nil, err
+		}
+		pushAuthConfig = authConfigs[registryServer]
```

Comment:
I think the `authConfigs` keys are host names (no https://), my own auth configs are keys like "ghcr.io" and "registry-1.docker.io".

Lines up with the other comment.

Reply:
My comment could be wrong. 

I dug a bit deeper into the implementation I'm most familiar with. There's some unusual behavior around the use of `docker.io`.

In the `oras` (OCI Registry as Storage) client, the hostname of an artifact/image reference such as `docker.io/library/hello-world` has a hardcoded mapping like so:

```go
// Host returns the host name of the registry.
func (r Reference) Host() string {
	if r.Registry == "docker.io" {
		return "registry-1.docker.io"
	}
	return r.Registry
}
```

And when querying the credential store, if the hostname is "registry-1.docker.io", the hostname is replaced with "https://index.docker.io/v1/":

```go
		store, err := credential.NewStore(opts.Configs...)
		if err != nil {
			return nil, err
		}
		// For a user case with a registry from 'docker.io', the hostname is "registry-1.docker.io"
		// According to the the behavior of Docker CLI,
		// credential under key "https://index.docker.io/v1/" should be provided
		if registry == "docker.io" {
			client.Credential = func(ctx context.Context, hostname string) (auth.Credential, error) {
				if hostname == "registry-1.docker.io" {
					hostname = "https://index.docker.io/v1/"
				}
				return store.Credential(ctx, hostname)
			}
		} else {
			client.Credential = store.Credential
		}
```

I suspect we'd need to do a mapping of our own - if the image name is `docker.io` we transparently map it to `registry-1.docker.io`, and if the registry is `registry-1.docker.io`, we the push auth config used is `https://index.docker.io/v1`.

Very confusing. I wonder what part(s) of this buildkit handles for us, though?

Reply:
The `oras` mapping reflects what the credential store type does. When we get the credentials from the config file, the call to `GetAllCredentials` returns a `map[string]clitypes.AuthConfig`, where the `string` is the full registry address including `https://`. I mapped this out via the `q` debugger. I can make this more clear with a better function name and some comments.
However, now that I think about it, this particular implementation puts the onus on the user to add the `https://` as the [failing Python test shows](https://github.com/pulumi/pulumi-docker/actions/runs/3991788221/jobs/6847039048#step:22:270)
...I think we can do a bit better and add it on the user's behalf; if it is still incorrect we can let Docker itself handle the error.

Reply:
That makes sense, thanks for talking that through.

> I think we can do a bit better and add it on the user's behalf

Definitely!